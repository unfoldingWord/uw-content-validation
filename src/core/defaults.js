export const DEFAULT_EXCERPT_LENGTH = 20;

export const REPO_CODES_LIST = [
  'UHB', 'UGNT',
  'LT', 'ST',
  'TA', 'TW', 'TWL',
  // The ones with 2 suffix are the 2021 new TSV format repos
  'TN', 'TN2', 'TQ', 'TQ2',
  'SN', 'SQ',
  'UGL', 'UHAL',
  'OBS', 'OBS-TWL',
  'OBS-TN', 'OBS-TN2', 'OBS-TQ', 'OBS-TQ2',
  'OBS-SN', 'OBS-SQ',
];
