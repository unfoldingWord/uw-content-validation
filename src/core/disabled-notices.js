// eslint-disable-next-line no-unused-vars
import { debugLog, userLog, functionLog } from './utilities';

/* This file handles the suppression of notices where we don’t want to disable or remove the actual check,
    but we just want to disable it for certain resources to handle special cases.
    In some cases, it’s to handle software deficiencies.

  Some examples:
    An article about punctuation discusses the use of "(" (which would normally cause a warning)
    An article about how to encode markdown discusses "#" (which would normally cause a warning)

    NOTE: This is only recommended for resources which are relatively stable,
          e.g., completed book packages
        as it can rely on details like filename and maybe line number.
*/


// const DISABLED_NOTICES_VERSION_STRING = '0.3.9';


const disabledNotices = [
  // Just enter enough details to disable the required message(s) and no unwanted ones
  //  i.e., you can be as specific as you need about username and/or lineNumber, etc.

  { repoCode: 'TN', priority: 450, }, // TN "Resource container link should have '*' language code with (not 'en')" disabled as tC can’t handle it yet!
  { repoCode: 'OBS-TN', priority: 450, }, // OBS-TN "Resource container link should have '*' language code with (not 'en')" disabled as tS can’t handle it yet!
  { repoCode: 'TW', priority: 450, }, // TW "Resource container link should have '*' language code with (not 'en')" disabled as tC can’t handle it yet!
  { extra: 'TW', priority: 450, }, // TW "Resource container link should have '*' language code with (not 'en')" disabled as tC can’t handle it yet!

  { repoCode: 'ST', message: "Bad punctuation nesting: } closing character doesn’t match", bookID: 'NEH', }, // 777 - complex { } nesting in direct speech
  { repoCode: 'ST', message: "Bad punctuation nesting: ” closing character doesn’t match", bookID: 'NEH', }, // 777 - complex { } nesting in direct speech

  { repoCode: 'LT', priority: 638, fieldName: '\\p', }, // "Only found whitespace" tC3 outputs trailing spaces here
  { repoCode: 'ST', priority: 638, fieldName: '\\p', }, // "Only found whitespace" tC3 outputs trailing spaces here
  { repoCode: 'LT', priority: 124, excerpt: '\\p␣␣', }, // "Unexpected double spaces" tC3 outputs trailing spaces here
  { repoCode: 'ST', priority: 124, excerpt: '\\p␣␣', }, // "Unexpected double spaces" tC3 outputs trailing spaces here
  { repoCode: 'LT', message: "Unexpected space after | character", }, // 192 tC3 outputs an unneeded/unwanted space in \zaln-s
  { repoCode: 'ST', message: "Unexpected space after | character", }, // 192 tC3 outputs an unneeded/unwanted space in \zaln-s
  { repoCode: 'LT', priority: 95, }, // "Unexpected trailing space(s)" tC3 outputs trailing spaces all over the place
  { repoCode: 'ST', priority: 95, }, // "Unexpected trailing space(s)" tC3 outputs trailing spaces all over the place

  { repoCode: 'TA', priority: 104, }, // "Unexpected trailing line break" UTA uses trailing <BR> for (1) (2) (3) style numbered lists

  { repoCode: 'TA', message: "Unexpected – character after space", }, // 191 -- temp en-dashes
  { repoCode: 'TA', message: "Unexpected space after – character", }, // 192 -- temp en-dashes

  { repoCode: 'TA', priority: 172, }, // "Header levels should only increment by one" not required for TA

  { repoCode: 'TA', filename: 'translate/translate-alphabet/01.md', message: "At end of text with unclosed ‘ opening character", }, // 768
  { repoCode: 'TA', filename: 'translate/translate-alphabet/01.md', message: "Mismatched ‘’ characters", }, // 462
  { repoCode: 'TA', filename: 'translate/translate-textvariants/01.md', message: "Unexpected space after [ character", }, // 192
  { repoCode: 'TA', filename: 'translate/translate-formatsignals/01.md', message: "Unexpected space after ( character", }, // 192

  { repoCode: 'TN', excerpt: ' brackets [ ] to ind', message: "Unexpected space after [ character", }, // 192

  // This file explains how to use markdown headings
  { repoCode: 'TA', filename: 'translate/file-formats/01.md', message: "Unexpected # character at start of line", }, // 195

  // This file has a space-separated list of typical quotation symbols
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected space after “ character", lineNumber: 3, }, // 192
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected space after “ character", lineNumber: 16, }, // 192
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected space after ‘ character", lineNumber: 16, }, // 192
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected space after « character", }, // 192
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected space after ‹ character", }, // 192
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected space after — character", }, // 192

  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected . character after space", lineNumber: 16, }, // 191
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected ” character after space", }, // 191
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected ’ character after space", }, // 191
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected » character after space", }, // 191
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected › character after space", }, // 191
  { repoCode: 'TA', filename: 'translate/figs-quotemarks/01.md', message: "Unexpected — character after space", }, // 191

  { repoCode: 'TA', filename: 'translate/figs-metaphor/01.md', message: "Possible misplaced ( character", }, // 17
];


/**
 *
 * @param {Object} givenNotice
 * @returns true if the givenNotice has a match in the disabledNotices list above
 */
export function isDisabledNotice(givenNotice) {
  // NOTE: The function will fail if repoCode is not set in the notices passed to this function
  // NOTE: 'extra' is still valid at this point (not yet prepended to 'message')
  // functionLog(`isDisabledNotice(${JSON.stringify(givenNotice)})…`);
  // if (givenNotice.repoCode === undefined) debugLog(`isDisabledNotice() cannot work without repoCode for ${JSON.stringify(givenNotice)}`);
  for (const disabledNotice of disabledNotices) {
    let matchedAllSpecifiedFields = true;
    for (const propertyName in disabledNotice)
      // if ((propertyName !== 'repoCode' || givenNotice.repoCode !== undefined) // Some individual checks don’t set repoCode
      // && (givenNotice[propertyName] !== disabledNotice[propertyName])) {
      if (givenNotice[propertyName] !== disabledNotice[propertyName]) {
        matchedAllSpecifiedFields = false;
        break;
      }
    if (matchedAllSpecifiedFields) {
      // debugLog(`  isDisabledNotice() returning true for ${JSON.stringify(disabledNotice)}`);
      return true;
    }
  }
  return false;
}


/**
 *
 * @param {Array} givenNoticeList
 * @returns a new list of notices with disabled ones removed
 */
export function removeDisabledNotices(givenNoticeList) {
  // NOTE: The function will fail if repoCode is not set in the notices passed to this function
  const remainingNoticeList = [];
  let givenRepoCodeNotice = false;
  for (const thisNotice of givenNoticeList) {
    if (thisNotice.repoCode === undefined && !givenRepoCodeNotice) { debugLog(`removeDisabledNotices() cannot work without repoCode for ${JSON.stringify(thisNotice)} in list of ${givenNoticeList.length} notices.`); givenRepoCodeNotice = true; }
    if (!isDisabledNotice(thisNotice))
      remainingNoticeList.push(thisNotice);
    // else userLog(`  Removing disabled ${JSON.stringify(thisNotice)}`);
  }
  if (givenNoticeList.length > 2 // This stops the msg from being displayed for lots of tests
    && remainingNoticeList.length !== givenNoticeList.length)
    userLog(`removeDisabledNotices() returning ${remainingNoticeList.length} out of ${givenNoticeList.length} notices`);
  return remainingNoticeList;
}
