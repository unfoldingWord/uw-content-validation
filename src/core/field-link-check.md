## Basic Link Field Check Sandbox

This function is for checking text fields that are links, or that contain links.

```js
// The control code in this box is editable for changing settings—
//    simply click inside here and add, change, or delete text as required.
// Note that (gray) lines starting with // are "comments", i.e., they are ignored by the software
//    so if you want to enable those lines, you must remove the // from the beginning of the line.

import { checkFieldLinks } from './field-link-check';
import { processNoticesToErrorsWarnings } from '../demos/notice-processing-functions';
import { RenderRawResults, RenderSuccessesErrorsWarnings } from '../demos/RenderProcessedResults';
import { userLog } from './utilities';

// Empty, space, link, RC, good, and bad text samples
const textE = "";
const textS = " ";
const textL = "https://unfoldingWord.org";
const textRC1 = "rc://en/obs/book/obs/02/03";
const textRC2 = "rc://en/tw/dict/bible/other/canaan";
const textMD1 = "Have a look [here](https://unfoldingWord.org/)!";
const textMD2 = " Look at [[https://unfoldingWord.org]] ";
const textMD3 = "Now  look at [[http://door43.org]] ";

// Easy for user to change text used in next line (to demonstrate differing results)
const username = 'unfoldingWord';
const chosenTextName = "textMD3";
const chosenText = textMD3;

let rawResults;

// Define our callback function
function acceptUpdatedResult(newLinkResult){
    // Update the results with later results
    userLog(`acceptUpdatedResult callback function is updating result now with ${newLinkResult.noticeList.length} notices.`);
    rawResults.noticeList = newLinkResult.noticeList;
    // Now how can we tell Styleguidist to refresh???
}

const linkOptions = {
    expectedCount: 1, // 0 = all links optional (no minimal number expected)
    allowedCount: 0, // 0 = no maximum limit
    otherTextAllowed: true, // if false, field is just the link only
    linkTypesAllowed: ['RC', 'md', 'naked',],
        // 'RC' (ResourceContainer) allows 'rc://...` links
        // 'md' allows '[this](http://here.org)' and [[http://here.org]]
        // 'naked' allows 'http:://example.org' and even 'example.org'
    checkTargets: true, // Attempt to retrieve the link(s) -- slows it down
    callbackFunction: acceptUpdatedResult, // Required to accept updated results if checking targets
};

// This function returns the results of the fast checks
//  and if specified in linkOptions, the callback will update result later with results of slower checks
rawResults = checkFieldLinks(username, 'en', '', chosenTextName, chosenText, linkOptions, 'that was supplied');
if (!rawResults.successList || !rawResults.successList.length)
    rawResults.successList = ["Done basic link checks"];

<>
<b>Check</b> "{chosenText}"<br/><br/>
<RenderRawResults results={rawResults} />
</>
```
