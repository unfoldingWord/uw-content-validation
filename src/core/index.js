export * from './books/';
export * from './defaults';
export * from './utilities';
export * from './getApi';
export * from './text-handling-functions';

export * from './field-text-check';
export * from './file-text-check';
export * from './lexicon-file-contents-check';
export * from './markdown-text-check';
export * from './markdown-file-contents-check';
export * from './manifest-text-check';
export * from './notes-tsv7-row-check';
export * from './notes-tsv7-table-check';
export * from './plain-text-check';
export * from './questions-tsv7-row-check';
export * from './questions-tsv7-table-check';
export * from './notes-links-check';
export * from './tn-tsv9-row-check';
export * from './tn-tsv9-table-check';
export * from './twl-tsv6-row-check';
export * from './twl-tsv6-table-check';
export * from './usfm-text-check';
export * from './yaml-text-check';
