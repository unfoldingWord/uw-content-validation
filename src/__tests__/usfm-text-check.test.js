/* eslint-env jest */

import { checkUSFMText } from '../core/usfm-text-check';
import Path from "path";
import fs from 'fs-extra';

const optionalCheckingOptions = {
  originalLanguageRepoUsername: 'unfoldingWord',
  taRepoUsername: 'unfoldingWord',
  disableAllLinkFetchingFlag: true, // until we can solve localforage error: No available storage method found
  // The following flags have no meaning if the above is set to true
  disableLinkedTAArticlesCheckFlag: true,
  disableLinkedTWArticlesCheckFlag: true,
  disableLexiconLinkFetchingFlag: true,
  disableLinkedLexiconEntriesCheckFlag: true,
  getFile: params => {
    const { username, repository, path } = params;
    // console.log(`usfm-text-check.test getFile(${username}, ${repository}, ${path})`)
    const filePath = Path.join('./src/__tests__/fixtures', username, repository, path);
    if (fs.existsSync(filePath)) {
      return fs.readFileSync(filePath).toString();
    }
    // eslint-disable-next-line no-throw-literal
    throw `usfm-text-check.test getFile(): Could not find ${filePath}`;
  }
}

describe('checkUSFMText() - ', () => {
  const username = 'unfoldingWord';
  const languageCode = 'en';
  const repoCode = 'LT';

  describe('short snippet tests - ', () => {
    it('should pass on smallest snippet', async () => {
      const usfmText = `\\id 2JN
\\usfm 3.0
\\ide UTF-8
\\mt1 2 John
\\c 1
\\p \\v 1 a b c\\v 2 b c d\\v 3 c d e\\v 4 d e f\\v 5 e f g\\v 6 f g h\\v 7 g h i\\v 8 h i j\\v 9 i j k\\v 10 j k l\\v 11 k l m\\v 12 l m n\\v 13 n o p
`; // There's a minimum number of words/characters expected in each verse
      const rawResults = await checkUSFMText(username, languageCode, repoCode, '2JN', 'test.usfm', usfmText, 'from test snippet', optionalCheckingOptions);
      // console.log(`checkUSFMText from usfmText='${usfmText}' got rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(0);
      expect(rawResults).toMatchSnapshot();
    });

    it('should fail on empty string', async () => {
      const rawResults = await checkUSFMText(username, languageCode, repoCode, 'RUT', 'test.usfm', "", 'from test snippet', optionalCheckingOptions);
      // console.log(`checkUSFMText rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(6);
      expect(rawResults).toMatchSnapshot();
    });

    it('should fail on non-USFM', async () => {
      const usfmText = "2:12\tgnn5\t\tfigs-parallelism\tשְׁלֵמָ֗ה\t1\tThis is a poetic expression that is very similar to the previous sentence. Alternate translation: “May Yahweh fully give to you everything that you deserve” (See: [[rc://*/ta/man/translate/figs-parallelism]], [Doublet](../figs-doublet/01.md))\n";
      const rawResults = await checkUSFMText(username, languageCode, repoCode, 'RUT', 'test.usfm', usfmText, 'from test snippet', optionalCheckingOptions);
      // console.log(`checkUSFMText from usfmText='${usfmText}' got rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(12);
      expect(rawResults).toMatchSnapshot();
    });

    it('should fail on no ID line', async () => {
      const rawResults = await checkUSFMText(username, languageCode, repoCode, 'RUT', 'test.usfm', "\\ide UTF-8\n", 'from test snippet', optionalCheckingOptions);
      // console.log(`checkUSFMText rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(6);
      expect(rawResults).toMatchSnapshot();
    });

    it('should fail on wrong book code', async () => {
      const rawResults = await checkUSFMText(username, languageCode, repoCode, 'RUT', 'test.usfm', "\\id GEN\n", 'from test snippet', optionalCheckingOptions);
      // console.log(`checkUSFMText rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(6);
      expect(rawResults).toMatchSnapshot();
    });

    it('should fail on gap between "5, 000"', async () => {
      const usfmText = `\\id 2JN
\\usfm 3.0
\\ide UTF-8
\\mt1 2 John
\\c 1
\\p \\v 1 a b c\\v 2 b c d\\v 3 c d e\\v 4 d e f\\v 5 e f g\\v 6 f g h\\v 7 g h i\\v 8 h i j\\v 9 i j k\\v 10 j k l\\v 11 k l m\\v 12 l m n
\\v 13 (\\zaln-s |x-strong="G10630" x-lemma="γάρ" x-morph="Gr,CC,,,,,,,," x-occurrence="1" x-occurrences="1" x-content="γὰρ"\\*\\w For|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G15100" x-lemma="εἰμί" x-morph="Gr,V,IIA3,,P," x-occurrence="1" x-occurrences="1" x-content="ἦσαν"\\*\\w there|x-occurrence="1" x-occurrences="1"\\w*
\\w were|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G56160" x-lemma="ὡσεί" x-morph="Gr,D,,,,,,,,," x-occurrence="1" x-occurrences="2" x-content="ὡσεὶ"\\*\\w about|x-occurrence="1" x-occurrences="2"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G40000" x-lemma="πεντακισχίλιοι" x-morph="Gr,EN,,,,NMP," x-occurrence="1" x-occurrences="1" x-content="πεντακισχίλιοι"\\*\\w 5|x-occurrence="1" x-occurrences="1"\\w*,
\\w 000|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G04350" x-lemma="ἀνήρ" x-morph="Gr,N,,,,,NMP," x-occurrence="1" x-occurrences="1" x-content="ἄνδρες"\\*\\w men|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*.)
`; // There's a minimum number of words/characters expected in each verse
      const rawResults = await checkUSFMText(username, languageCode, repoCode, '2JN', 'test.usfm', usfmText, 'from test snippet', optionalCheckingOptions);
      // console.log(`checkUSFMText from usfmText='${usfmText}' got rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(2);
      expect(rawResults).toMatchSnapshot();
    });

    it('should not fail on gap between "5, 000"', async () => {
      const usfmText = `\\id 2JN
\\usfm 3.0
\\ide UTF-8
\\mt1 2 John
\\c 1
\\p \\v 1 a b c\\v 2 b c d\\v 3 c d e\\v 4 d e f\\v 5 e f g\\v 6 f g h\\v 7 g h i\\v 8 h i j\\v 9 i j k\\v 10 j k l\\v 11 k l m\\v 12 l m n
\\v 13 (\\zaln-s |x-strong="G10630" x-lemma="γάρ" x-morph="Gr,CC,,,,,,,," x-occurrence="1" x-occurrences="1" x-content="γὰρ"\\*\\w For|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G15100" x-lemma="εἰμί" x-morph="Gr,V,IIA3,,P," x-occurrence="1" x-occurrences="1" x-content="ἦσαν"\\*\\w there|x-occurrence="1" x-occurrences="1"\\w*
\\w were|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G56160" x-lemma="ὡσεί" x-morph="Gr,D,,,,,,,,," x-occurrence="1" x-occurrences="2" x-content="ὡσεὶ"\\*\\w about|x-occurrence="1" x-occurrences="2"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G40000" x-lemma="πεντακισχίλιοι" x-morph="Gr,EN,,,,NMP," x-occurrence="1" x-occurrences="1" x-content="πεντακισχίλιοι"\\*\\w 5|x-occurrence="1" x-occurrences="1"\\w*,\\w 000|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G04350" x-lemma="ἀνήρ" x-morph="Gr,N,,,,,NMP," x-occurrence="1" x-occurrences="1" x-content="ἄνδρες"\\*\\w men|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*.)
`; // There's a minimum number of words/characters expected in each verse
      const rawResults = await checkUSFMText(username, languageCode, repoCode, '2JN', 'test.usfm', usfmText, 'from test snippet', optionalCheckingOptions);
      // console.log(`checkUSFMText from usfmText='${usfmText}' got rawResults=${JSON.stringify(rawResults)}`);
      expect(rawResults.noticeList.length).toEqual(1); // TODO: Why do we get "verse number didn't increment correctly"???
      expect(rawResults).toMatchSnapshot();
    });

  });

});
