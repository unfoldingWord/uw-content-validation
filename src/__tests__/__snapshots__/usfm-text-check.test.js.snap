// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkUSFMText() -  short snippet tests -  should fail on empty string 1`] = `
Object {
  "noticeList": Array [
    Object {
      "bookID": "RUT",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "USFM file must start with a valid \\\\id line",
      "priority": 994,
    },
    Object {
      "bookID": "RUT",
      "filename": "test.usfm",
      "lineNumber": 2,
      "location": " from test snippet",
      "message": "USFM file is recommended to have \\\\ide line",
      "priority": 719,
    },
    Object {
      "bookID": "RUT",
      "excerpt": "missing \\\\usfm",
      "filename": "test.usfm",
      "location": " from test snippet",
      "message": "Missing expected USFM line",
      "priority": 519,
    },
    Object {
      "bookID": "RUT",
      "excerpt": "missing \\\\mt1",
      "filename": "test.usfm",
      "location": " from test snippet",
      "message": "Missing expected USFM line",
      "priority": 921,
    },
    Object {
      "bookID": "RUT",
      "filename": "text extracted from test.usfm",
      "location": " from test snippet",
      "message": "Only found whitespace",
      "priority": 638,
    },
    Object {
      "bookID": "RUT",
      "characterIndex": 0,
      "details": "
> 1 | 
      ^
Expected \\"\\\\\\\\\\"",
      "excerpt": "",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "USFMGrammar check failed",
      "priority": 840,
    },
  ],
  "successList": Array [
    "Checked all 1 line for RUT from test snippet",
    "Checked C:V patterns for RUT from test snippet",
    "Checked USFM Grammar (strict mode)  (but the USFM DIDN’T validate)",
  ],
}
`;

exports[`checkUSFMText() -  short snippet tests -  should fail on gap between "5, 000" 1`] = `
Object {
  "noticeList": Array [
    Object {
      "C": "1",
      "V": "13",
      "bookID": "2JN",
      "characterIndex": 3,
      "excerpt": "13 (\\\\zaln-… (1 → 13)",
      "filename": "test.usfm",
      "lineNumber": 7,
      "location": " from test snippet",
      "message": "Verse number didn’t increment correctly",
      "priority": 763,
    },
    Object {
      "bookID": "2JN",
      "excerpt": "5, 000",
      "filename": "text extracted from test.usfm",
      "location": " from test snippet",
      "message": "Found separated digits",
      "priority": 498,
    },
  ],
  "successList": Array [
    "Checked all 14 lines for 2JN from test snippet",
    "Checked C:V patterns for 2JN from test snippet",
    "Checked USFM Grammar (strict mode) without errors",
  ],
}
`;

exports[`checkUSFMText() -  short snippet tests -  should fail on no ID line 1`] = `
Object {
  "noticeList": Array [
    Object {
      "bookID": "RUT",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "USFM file must start with a valid \\\\id line",
      "priority": 994,
    },
    Object {
      "bookID": "RUT",
      "filename": "test.usfm",
      "lineNumber": 2,
      "location": " from test snippet",
      "message": "USFM file is recommended to have \\\\ide line",
      "priority": 719,
    },
    Object {
      "bookID": "RUT",
      "excerpt": "missing \\\\usfm",
      "filename": "test.usfm",
      "location": " from test snippet",
      "message": "Missing expected USFM line",
      "priority": 519,
    },
    Object {
      "bookID": "RUT",
      "excerpt": "missing \\\\mt1",
      "filename": "test.usfm",
      "location": " from test snippet",
      "message": "Missing expected USFM line",
      "priority": 921,
    },
    Object {
      "bookID": "RUT",
      "filename": "text extracted from test.usfm",
      "location": " from test snippet",
      "message": "Only found whitespace",
      "priority": 638,
    },
    Object {
      "bookID": "RUT",
      "characterIndex": 3,
      "details": "
> 1 | \\\\ide UTF-8
         ^
  2 | 
Expected \\" \\"",
      "excerpt": "\\\\ide UTF-8",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "USFMGrammar check failed",
      "priority": 840,
    },
  ],
  "successList": Array [
    "Checked all 2 lines for RUT from test snippet",
    "Checked C:V patterns for RUT from test snippet",
    "Checked USFM Grammar (strict mode)  (but the USFM DIDN’T validate)",
  ],
}
`;

exports[`checkUSFMText() -  short snippet tests -  should fail on non-USFM 1`] = `
Object {
  "noticeList": Array [
    Object {
      "bookID": "RUT",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "USFM file must start with a valid \\\\id line",
      "priority": 994,
    },
    Object {
      "bookID": "RUT",
      "filename": "test.usfm",
      "lineNumber": 2,
      "location": " from test snippet",
      "message": "USFM file is recommended to have \\\\ide line",
      "priority": 719,
    },
    Object {
      "C": "0",
      "V": "1",
      "bookID": "RUT",
      "characterIndex": 0,
      "excerpt": "2",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Expected line to start with backslash",
      "priority": 880,
    },
    Object {
      "C": "0",
      "V": "1",
      "bookID": "RUT",
      "characterIndex": 208,
      "excerpt": "…See: [[rc://*/ta/man…",
      "fieldName": "\\\\NONE",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Unexpected doubled / characters",
      "priority": 177,
    },
    Object {
      "C": "0",
      "V": "1",
      "bookID": "RUT",
      "characterIndex": 203,
      "excerpt": "…ve” (See: [[rc://*/t…",
      "fieldName": "\\\\NONE",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Unexpected doubled [ characters",
      "priority": 177,
    },
    Object {
      "C": "0",
      "V": "1",
      "bookID": "RUT",
      "characterIndex": 245,
      "excerpt": "…arallelism]], [Doubl…",
      "fieldName": "\\\\NONE",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Unexpected doubled ] characters",
      "priority": 177,
    },
    Object {
      "C": "0",
      "V": "1",
      "bookID": "RUT",
      "characterIndex": 259,
      "details": "might be intended to be … (ellipse)???",
      "excerpt": "…[Doublet](../figs-do…",
      "fieldName": "\\\\NONE",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Unexpected doubled . characters",
      "priority": 177,
    },
    Object {
      "C": "0",
      "V": "1",
      "bookID": "RUT",
      "characterIndex": 280,
      "excerpt": "…blet/01.md))",
      "fieldName": "\\\\NONE",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Unexpected doubled ) characters",
      "priority": 177,
    },
    Object {
      "C": "0",
      "V": "1",
      "bookID": "RUT",
      "characterIndex": 207,
      "excerpt": "…(See: [[rc://*/ta/ma…",
      "fieldName": "\\\\NONE",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Unexpected link",
      "priority": 765,
    },
    Object {
      "bookID": "RUT",
      "excerpt": "missing \\\\usfm",
      "filename": "test.usfm",
      "location": " from test snippet",
      "message": "Missing expected USFM line",
      "priority": 519,
    },
    Object {
      "bookID": "RUT",
      "excerpt": "missing \\\\mt1",
      "filename": "test.usfm",
      "location": " from test snippet",
      "message": "Missing expected USFM line",
      "priority": 921,
    },
    Object {
      "bookID": "RUT",
      "characterIndex": 0,
      "details": "
> 1 | 2:12	gnn5		figs-parallelism	שְׁלֵמָ֗ה	1	This is a poetic expression that is very similar to the previous sentence. Alternate translation: “May Yahweh fully give to you everything that you deserve” (See: [[rc://*/ta/man/translate/figs-parallelism]], [Doublet](../figs-doublet/01.md))
      ^
  2 | 
Expected \\"\\\\\\\\\\"",
      "excerpt": "2:12	gnn5	…",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "USFMGrammar check failed",
      "priority": 840,
    },
  ],
  "successList": Array [
    "Checked all 2 lines for RUT from test snippet",
    "Checked C:V patterns for RUT from test snippet",
    "Checked USFM Grammar (strict mode)  (but the USFM DIDN’T validate)",
  ],
}
`;

exports[`checkUSFMText() -  short snippet tests -  should fail on wrong book code 1`] = `
Object {
  "noticeList": Array [
    Object {
      "bookID": "RUT",
      "filename": "test.usfm",
      "lineNumber": 2,
      "location": " from test snippet",
      "message": "USFM file is recommended to have \\\\ide line",
      "priority": 719,
    },
    Object {
      "C": "0",
      "V": "1",
      "bookID": "RUT",
      "characterIndex": 4,
      "details": "expected bookID='RUT'",
      "excerpt": "GEN",
      "filename": "test.usfm",
      "lineNumber": 1,
      "location": " from test snippet",
      "message": "Expected USFM \\\\id line to start with book identifier",
      "priority": 987,
    },
    Object {
      "bookID": "RUT",
      "excerpt": "missing \\\\usfm",
      "filename": "test.usfm",
      "location": " from test snippet",
      "message": "Missing expected USFM line",
      "priority": 519,
    },
    Object {
      "bookID": "RUT",
      "excerpt": "missing \\\\mt1",
      "filename": "test.usfm",
      "location": " from test snippet",
      "message": "Missing expected USFM line",
      "priority": 921,
    },
    Object {
      "bookID": "RUT",
      "filename": "text extracted from test.usfm",
      "location": " from test snippet",
      "message": "Only found whitespace",
      "priority": 638,
    },
    Object {
      "bookID": "RUT",
      "characterIndex": 0,
      "details": "
  1 | \\\\id GEN
> 2 | 
      ^
Expected \\"\\\\\\\\\\"",
      "excerpt": "",
      "filename": "test.usfm",
      "lineNumber": 2,
      "location": " from test snippet",
      "message": "USFMGrammar check failed",
      "priority": 840,
    },
  ],
  "successList": Array [
    "Checked all 2 lines for RUT from test snippet",
    "Checked C:V patterns for RUT from test snippet",
    "Checked USFM Grammar (strict mode)  (but the USFM DIDN’T validate)",
  ],
}
`;

exports[`checkUSFMText() -  short snippet tests -  should not fail on gap between "5, 000" 1`] = `
Object {
  "noticeList": Array [
    Object {
      "C": "1",
      "V": "13",
      "bookID": "2JN",
      "characterIndex": 3,
      "excerpt": "13 (\\\\zaln-… (1 → 13)",
      "filename": "test.usfm",
      "lineNumber": 7,
      "location": " from test snippet",
      "message": "Verse number didn’t increment correctly",
      "priority": 763,
    },
  ],
  "successList": Array [
    "Checked all 13 lines for 2JN from test snippet",
    "Checked C:V patterns for 2JN from test snippet",
    "Checked USFM Grammar (strict mode) without errors",
  ],
}
`;

exports[`checkUSFMText() -  short snippet tests -  should pass on smallest snippet 1`] = `
Object {
  "noticeList": Array [],
  "successList": Array [
    "Checked all 7 lines for 2JN from test snippet",
    "Checked C:V patterns for 2JN from test snippet",
    "Checked USFM Grammar (strict mode) without errors",
  ],
}
`;
